src/steps/generic.rs

```diff
@@ -1277,20 +1275,91 @@ pub fn run_uv(ctx: &ExecutionContext) -> Result<()> {
+            .trim_start_matches(' ');
+        // remove the tailing part " (c4d0caaee 2024-12-19)\n"
+        let first_whitespace_index = start_trimmed
+            .find(' ')
+            .expect("the output of `uv --version` changed, please file an issue to Topgrade");
+        // this should be our version str "0.5.11"
+        &start_trimmed[..first_whitespace_index]
+    };
+    let version =
+        Version::parse(version_str).expect("the output of `uv --version` changed, please file an issue to Topgrade");
+
+    if version < Version::new(0, 4, 25) {
+        // For uv before version 0.4.25 (exclusive), the `self` sub-command only
+        // exists under the `self-update` feature, we run `uv self --help` to check
+        // the feature gate.
+        let self_update_feature_enabled = ctx
+            .run_type()
+            .execute(&uv_exec)
+            .args(["self", "--help"])
+            .output_checked()
+            .is_ok();
 
-    if result.is_ok() {
-        ctx.run_type()
+        if self_update_feature_enabled {
+            ctx.run_type()
+                .execute(&uv_exec)
+                .args(["self", "update"])
+                .status_checked()?;
+        }
+    } else {
+        // After 0.4.25 (inclusive), running `uv self` succeeds regardless of the
+        // feature gate, so the above approach won't work.
+        //
+        // We run `uv self update` directly, if it outputs:
+        //
+        // "uv was installed through an external package manager, and self-update is not available. Please use your package manager to update uv.\n"
+
+        const ERROR_MSG: &str = "uv was installed through an external package manager, and self-update is not available. Please use your package manager to update uv.";
+
+        let output = ctx
+            .run_type()
             .execute(&uv_exec)
             .args(["self", "update"])
-            .status_checked()?;
-    }
+            // `output()` captures the output so that users won't see it for now.
+            .output()
+            .expect("this should be ok regardless of this child process's exit code");
+        let output = match output {
+            ExecutorOutput::Wet(wet) => wet,
+            ExecutorOutput::Dry => unreachable!("the whole function returns when we run `uv --version` under dry-run"),
+        };
+        let stderr = std::str::from_utf8(&output.stderr).expect("output should be UTF-8 encoded");
+
+        if stderr.contains(ERROR_MSG) {
+            // Feature `self-update` is disabled, nothing to do.
+        } else {
+            // Feature is enabled, flush the captured output so that users know we did the self-update.
+
+            std::io::stdout().write_all(&output.stdout)?;
+            std::io::stderr().write_all(&output.stderr)?;
+
+            // And, if self update failed, fail the step as well.
+            if !output.status.success() {
+                return Err(eyre!("uv self update failed"));
+            }
```

Comment:
Instead of raising a custom error, can we call `output.status_checked()?`?
```suggestion
            output.status_checked()?;
```

Reply:
Looks like no, `CommandExt` is only implemented for `std::process::Command` and `Executor`, `output` is of type `std::process::Output`

Reply:
Okay, that makes sense